File: sdk/dotnet/src/FormSpecLib/Ctx.cs
namespace FormSpec;

/// <summary>
/// The ctx.* surface handed to handlers. Every method call is an HTTP call
/// back to formspec-sidecar (§4.3) — the same primitive contract Starlark scripts use.
/// </summary>
public sealed class Ctx
{
    private readonly SidecarClient _client;

    internal Ctx(SidecarClient client) => _client = client;

    public CtxPrimitive Db() => new(_client, "db");
    public CtxPrimitive Cache() => new(_client, "cache");
    public CtxPrimitive Lock() => new(_client, "lock");
    public CtxPrimitive Queue() => new(_client, "queue");
    public CtxPrimitive PubSub() => new(_client, "pubsub");
    public CtxPrimitive Storage() => new(_client, "storage");
    public CtxPrimitive KvStore() => new(_client, "kvstore");

    /// <summary>Entity primitive — access entity records via named("module/entity").</summary>
    public CtxPrimitive Entity() => new(_client, "entity");
}

File: sdk/ruby/lib/formspec/ctx.rb
module FormSpec
  # The ctx.* surface handed to handlers. Every method call is an HTTP call
  # back to formspec-sidecar (§4.3) — the same primitive contract Starlark
  # scripts use.
  class Ctx
    def initialize(client)
      @client = client
    end

    def db      = CtxPrimitive.new(@client, "db")
    def cache   = CtxPrimitive.new(@client, "cache")
    def lock    = CtxPrimitive.new(@client, "lock")
    def queue   = CtxPrimitive.new(@client, "queue")
    def pubsub  = CtxPrimitive.new(@client, "pubsub")
    def storage = CtxPrimitive.new(@client, "storage")
    def kvstore = CtxPrimitive.new(@client, "kvstore")

    # Entity primitive — access entity records via named('module/entity').
    def entity = CtxPrimitive.new(@client, "entity")
  end
end
